src/api/app/queries/packages_finder.rb

```diff
@@ -0,0 +1,84 @@
+class PackagesFinder
+  def initialize(relation = Package.all)
+    @relation = relation
+  end
+
+  def by_package_and_project(package, project)
+    @relation.where(name: package, projects: { name: project }).includes(:project)
+  end
+
+  def find_by_attribute_type(attrib_type, package = nil)
+    # One sql statement is faster than a ruby loop
+    # attribute match in package or project
+    sql = build_sql_find_by_attribute(package)
+    finder_arguments = if package
+                         [sql, attrib_type.id.to_s, attrib_type.id.to_s, package]
+                       else
+                         [sql, attrib_type.id.to_s, attrib_type.id.to_s]
+                       end
+    find_package(finder_arguments)
+  end
+
+  def find_by_attribute_type_and_value(attrib_type, value, package = nil)
+    # One sql statement is faster than a ruby loop
+    sql = build_sql_find_by_attribute_and_value(package)
+    finder_arguments = if package
+                         [sql, attrib_type.id.to_s, value.to_s, package]
+                       else
+                         [sql, attrib_type.id.to_s, value.to_s]
+                       end
+    find_package(finder_arguments)
+  end
+
+  def forbidden_packages
+    @relation.where(project_id: Relationship.forbidden_project_ids)
+  end
+
+  def dirty_backend_packages
+    @relation.joins('left outer join backend_packages on backend_packages.package_id = packages.id').
+      where('backend_packages.package_id is null')
+  end
+
+  private
+
+  def find_package(args)
+    Package.find_by_sql(args).keep_if { |p| Package.check_access?(p) }
```

Comment:
Could you change `p` by something more explicit like `package`?

Reply:
we talked off-line.

Reply:
I don't see a problem to have single letter for some short block. i.e no issue with `0.upto(10).each { |i| i * 10 }` so if it's ok I would keep like that.